### PR TITLE
codegen: Fix typedef re-export in namespaces when bindings aren't at the root.

### PR DESCRIPTION
**Pull request by emilio**

### CODE REVIEW

**Comment by emilio** [comment]
Fixes #433

**Comment by fitzgen** [comment]
Can you also add this test case:

```c++
namespace JS {
class Value;
}
typedef enum JSWhyMagic {} JSWhyMagic;
namespace JS {
class Value {
public:
  void a(JSWhyMagic);
};
}
```

**Comment by emilio** [comment]
That's the exact test case that is in tree already (for #410)

**Comment by fitzgen** [comment]
> That's the exact test case that is in tree already (for #410)

:-/

When I try and compile the bindings generated from this test case I get:

```
$ rustc --crate-type lib js.rs
error[E0412]: cannot find type `JSWhyMagic` in module `root::JS`
  --> js.rs:22:34
   |
22 |                            arg1: root::JS::JSWhyMagic);
   |                                  ^^^^^^^^^^^^^^^^^^^^ not found in `root::JS`
   |
   = help: possible candidate is found in another module, you can import it into scope:
             `use root::JSWhyMagic;`

error[E0412]: cannot find type `JSWhyMagic` in module `root::JS`
  --> js.rs:29:46
   |
29 |             pub unsafe fn a(&mut self, arg1: root::JS::JSWhyMagic) {
   |                                              ^^^^^^^^^^^^^^^^^^^^ not found in `root::JS`
   |
   = help: possible candidate is found in another module, you can import it into scope:
             `use root::JSWhyMagic;`

error: aborting due to 2 previous errors
```

**Comment by fitzgen** [comment]
I am getting this same error in the compilation of the SpiderMonkey bindings

**Comment by emilio** [comment]
In tree it happens to compile because `root` is at the top of the module.

File: src/codegen/mod.rs [hunk 1]
```diff
@@ -23,36 +23,47 @@ use std::cell::Cell;
...
+}
+
+fn top_level_module_path(ctx: &BindgenContext, item: &Item) -> Vec<ast::Ident> {
+    let mut path = vec![ctx.rust_ident_raw("self")];
+
+    let super_ = ctx.rust_ident_raw("super");
+
+    for _ in 0..root_import_depth(ctx, item) {
+        path.push(super_.clone());
+    }
```
Comment (fitzgen) [inline src/codegen/mod.rs]: This whole function body could be

```rust
let super_ = ctx.rust_ident_raw("super");
let root = ctx.root_module().canonical_name(ctx);
let root_ident = ctx.rust_ident(&root);

iter::once(ctx.rust_ident_raw("self"))
    .chain(iter::repeat(super_.clone()).take(root_import_depth(ctx, item)))
    .chain(iter::once(root_ident))
    .collect()
```

**Pull request merged**